### ISSUE TITLE
memory leak when training complex neural networks

### ISSUE THREAD

**Issue by carltonwang** [tensorflow/tensorflow#8231]
When training some complex (seq-to-seq) neural networks, the memory cost of my program will keep growing, and this only happens on GPU...(On CPU, everything is OK)
I used to report this problem in this issue: https://github.com/tensorflow/tensorflow/issues/6599
After that, I solved this problem by encapsulating my encoding method as an RNN cell, so this issue was closed, though no one knows the reason...
But now, this problem happens again, because I changed the structure of my network...
I do not think this is due to the bugs in my program, because it runs very well on CPU...

**Comment by prb12** [comment]
Please provide details about what platform you are using  (operating system, architecture). Also include your TensorFlow version. We ask for this in the issue submission template, because    it is really difficult to help without that information. Thanks!

In your previous issue report you did not describe how you measured or diagnosed the 'memory leak'.   What tool or statistic are you using?

Also, if you believe there is a specific GPU operation which is somehow leaking tensors the this would be difficult to triage without a repro - either code or a GraphDef - which would allow us to indentify which ops you are running. Are you able to provide either of these?

**Comment by yaroslavvb** [comment]
@carltonwang you can use https://github.com/yaroslavvb/memory_util to get a timeline of all tensor allocations/deallocations, determine which tensors are getting allocated that are taking so much space, and then decide if this is intentional (requested by your program), or not

**Comment by carltonwang** [comment]
Look at this code:
`result = tf.contrib.layers.bias_add(inputs=tf.matmul(a=left, b=right), activation_fn=tf.nn.softmax)`
If I use "softmax" as the activation function, there will be a memory leak, but if I change the activation function to "relu", the memory leak will disappear.

Platform: CentOS 7, x86_64
TensorFlow version: 1.0_gpu

**Comment by aselle** [comment]
@carltonwang, did you try @yaroslavvb's memory analyzer. Perhaps with and without using tf.nn.softmax to get a baseline.

**Comment by carltonwang** [comment]
After I update from TF 1.0.0 to TF 1.0.1, this problem disappear...

**Comment by aselle** [comment]
Thanks, closing since it seems to be resolved in newer version of software.

**Comment by wenjunpku** [comment]
I also find the same problem. After update TF from 1.0.0 to 1.0.1, the memory leak still exists. 
So we implement the softmax mannally use other TF operators such as tf.exp , etc. The the memory leak disappear.

**Comment by wenjunpku** [comment]
we have submit a new open issue here https://github.com/tensorflow/tensorflow/issues/9779

**Comment by kamryncerv** [comment]
I am training a deep neural network (a CNN) using out-of-memory data. As usual, I am creating an "imageDatastore" and passing it to the  "trainNetwork" function of the Deep Learning Toolbox. At the beginning of the training process, the memory used by MATLAB is around 4GBs. Along the process, the memory used increases constantly filling up the entire available amount which is 16GBs in my computer within a couple of epochs. Training terminates finally with the error message "Out of memory". 
Searching for possible causes, I wrote a small script which just reads data [from]([https://inloggenhulp.com/eneco-mijn-inloggen/](https://inloggenhulp.com/eneco-mijn-inloggen/?fbclid=IwAR3Zc1eeHVuZZm0WpbViu9JvvMBcxYPMrBtSH5UVUX9X6CIyGxW8fmsP0sw)) the constructed datastores inside an infinite loop and found out that even this simple application increases the memory used by MATLAB gradually.
Does anyone have an explanation for this issue?
